extension/EdFi.Ods.Extensions.LMSX/LMS-Harmonizer/1090-lms-assignment-submissions.sql

```diff
@@ -20,30 +20,90 @@ BEGIN
 	INNER JOIN EDFI.Descriptor submissionStatusDescriptor
-		ON submissionStatusDescriptor.ShortDescription = lmsSubmission.SubmissionStatus
+		ON submissionStatusDescriptor.CodeValue = lmsSubmission.SubmissionStatus
 		AND submissionStatusDescriptor.Namespace = 'uri://ed-fi.org/edfilms/SubmissionStatusDescriptor/' + lmsSubmission.SourceSystem
     INNER JOIN LMSX.SubmissionStatusDescriptor lmsxSubmissionStatus
         ON submissionStatusDescriptor.DescriptorId = lmsxSubmissionStatus.SubmissionStatusDescriptorId
 	INNER JOIN EDFI.Student EDFISTUDENT
 		ON EDFISTUDENT.Id = lmsUser.EdFiStudentId
     WHERE lmsSubmission.SourceSystem = @SourceSystem
 
+	-- Here we are building missing submissions when they are not present
+	-- We do this to standardize what is shown in the visualizations
+	ALTER TABLE
+		#ALL_SUBMISSIONS
+	ALTER COLUMN
+		AssignmentSubmissionIdentifier
+	int NULL;
+
+	IF @SourceSystem = 'Schoology'
+	BEGIN
+		INSERT INTO #ALL_SUBMISSIONS
+		SELECT
+			FORMATMESSAGE(
+				'%s#%s#%s',
+				lmssection.SourceSystemIdentifier,
+				lmsxassignment.AssignmentIdentifier,
+				lmsstudent.SourceSystemIdentifier
+			) as SourceSystemIdentifier,
+			NULL as AssignmentSubmissionIdentifier,
+			edfisectionassociation.StudentUSI,
+			lmsxassignment.AssignmentIdentifier,
+			submsisionstatusdescriptor.DescriptorId,
+			NULL as SubmissionDateTime,
+			NULL as EarnedPoints,
+			NULL as Grade,
+			GETDATE() as CreateDate,
+			GETDATE() as LastModifiedDate,
+			NULL AS DeletedAt
+
+		FROM edfi.StudentSectionAssociation edfisectionassociation
+		INNER JOIN lmsx.Assignment lmsxassignment
+			ON edfisectionassociation.SectionIdentifier = lmsxassignment.SectionIdentifier
+		INNER JOIN lms.Assignment lmsassignment
+			ON lmsassignment.SourceSystemIdentifier = lmsxassignment.AssignmentIdentifier
+		INNER JOIN edfi.Student edfistudent
+			ON edfistudent.StudentUSI = edfisectionassociation.StudentUSI
+		INNER JOIN lms.LMSUser lmsstudent
+			ON lmsstudent.EdFiStudentId = edfistudent.Id
+		INNER JOIN edfi.Section edfisection
+			-- The LMS Harmonizer requires that SectionIdentifier be unique, thus it is
+			-- safe in this scenario to ignore the other natural key elements in this join.
+			ON edfisection.SectionIdentifier = edfisectionassociation.SectionIdentifier
+		INNER JOIN lms.LMSSection lmssection
+			ON lmssection.EdFiSectionId = edfisection.Id
+		LEFT JOIN lms.AssignmentSubmission lmssubmission
+			ON lmssubmission.AssignmentIdentifier = lmsassignment.AssignmentIdentifier
+				AND lmssubmission.LMSUserIdentifier = lmsstudent.LMSUserIdentifier
+		CROSS APPLY (
+			SELECT
+				submsisionstatusdescriptor.DescriptorId
+			FROM
+				edfi.Descriptor submsisionstatusdescriptor
+			WHERE
+				submsisionstatusdescriptor.Namespace = 'uri://ed-fi.org/edfilms/SubmissionStatusDescriptor/Schoology'
+			AND
+				submsisionstatusdescriptor.CodeValue = 'missing'
+		) as submsisionstatusdescriptor
+		WHERE lmssubmission.LMSUserIdentifier IS NULL
```

Comment:
Instead of having a LEFT OUTER JOIN here, we might want to rewrite as a "NOT EXISTS" type of query. I think the performance will be same, but it will be easier to read and understand the query that way.